library/src/main/java/com/datatorrent/lib/state/managed/TimeBucketAssigner.java

```diff
@@ -0,0 +1,223 @@
+ * is expired. These boundaries slide between application window by another thread and not the operator thread.
+ */
+public class TimeBucketAssigner implements Component<Context.OperatorContext>
+{
+  @NotNull
+  @FieldSerializer.Bind(JavaSerializer.class)
+  private Duration expireBefore = Duration.standardDays(2);
+
+  @FieldSerializer.Bind(JavaSerializer.class)
+  private Duration bucketSpan;
+
+  private long bucketSpanMillis;
+
+  private long fixedStartTime;
+  private long startTime;
+  private long endTime;
+  private int numBuckets;
+
+  private boolean initialized;
+
+  private transient WindowBoundedService windowBoundedService;
+
+  @NotNull
+  private final transient Set<Listener> listeners = Sets.newHashSet();
+
+  private final transient Runnable expiryTask = new Runnable()
+  {
+    @Override
+    public void run()
+    {
+      synchronized (lock) {
+        startTime += bucketSpanMillis;
+        endTime += bucketSpanMillis;
+        for (Listener listener : listeners) {
+          listener.purgeTimeBucketsBefore(startTime);
+        }
+      }
+    }
+  };
+
+  private final transient Object lock = new Object();
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    if (!initialized) {
+      if (bucketSpan == null) {
+        bucketSpan = Duration.millis(context.getValue(Context.OperatorContext.APPLICATION_WINDOW_COUNT) *
+            context.getValue(Context.DAGContext.STREAMING_WINDOW_SIZE_MILLIS));
+      }
+      Calendar calendar = Calendar.getInstance();
+      long now = calendar.getTimeInMillis();
+      fixedStartTime = now - expireBefore.getMillis();
+      startTime = fixedStartTime;
+
+      bucketSpanMillis = bucketSpan.getMillis();
+      numBuckets = (int)((expireBefore.getMillis() + bucketSpanMillis - 1) / bucketSpanMillis);
+      endTime = startTime + (numBuckets * bucketSpanMillis);
+
+      initialized = true;
+    }
+    windowBoundedService = new WindowBoundedService(bucketSpanMillis, expiryTask);
+    windowBoundedService.setup(context);
+  }
+
+  public void beginWindow(long l)
+  {
+    windowBoundedService.beginWindow(l);
+  }
+
+  public void endWindow()
+  {
+    windowBoundedService.endWindow();
+  }
+
+  /**
+   * Get the bucket key for the long value.
+   *
+   * @param value value from which bucket key is derived.
+   * @return -1 if value is already expired; bucket key otherwise.
+   */
+  public long getTimeBucketFor(long value)
+  {
+    long lstart;
+    long lend;
+    synchronized (lock) {
+      lstart = startTime;
+      lend = endTime;
+    }
+    if (value < lstart) {
+      return -1;
+    }
+    long diffFromStart = value - fixedStartTime;
+    long key = diffFromStart / bucketSpanMillis;
+    if (value > lend) {
+      long move = ((value - lend) / bucketSpanMillis + 1) * bucketSpanMillis;
+      synchronized (lock) {
+        startTime = lstart + move;
+        endTime = lend + move;
+      }
```

Comment:
>> Can (value - lend) ever exceed bucketSpanMillis ?
Yes. ```startTime``` is initialized during setup() and ```endTime = startTime + numBucket*bucketSpan```. 

However ```value```  can come from the event. This time in the event can be ahead of ```endTime``` (we have seen this in a real application).  This triggers the whole boundary to move ahead.

>> Second, it seems like this assignment could overwrite the assignment made to startTime and endTime by the expiryTask thread in some scenarios, for example: (a) We copy the values in the first synchronized block above (b) The expiryTask thread increments values by bucketSpanMillis (c) We overwrite those values by the values computed here. Is this OK ?

We haven't seen a problem so far with this in BucketManager because the time in the event is given a priority over expiry task. However here there maybe an issue because the listener (```BucketsDataManager```) is notified to purge data before start time.

Will club (a) and (c) in one block